Run EvoSuite in new process

gandreadis opened the pull request:
This PR adds functionality to launch EvoSuite in a new process and let it generate tests for the class it is pointed to. I've added disabled its CSV statistics output and added options for us to configure whether we want to see its output logged to `System.out` (turned off by default, for now).

Unfortunately, I was not able to launch EvoSuite directly from its Java API, as it left a number of threads running in the background after termination (resorting to a `System.exit` call to terminate them). I've filed an [issue report](https://github.com/EvoSuite/evosuite/issues/194) with the EvoSuite team to check whether they have any suggestions on this.

casperboone commented on subprojects/test-generator/src/test/kotlin/org/cafejojo/schaapi/testgenerator/EvoSuiteRunnerTest.kt:
```diff
@@ -0,0 +1,40 @@
+    val classPath = EvoSuiteRunnerTest::class.java.getResource("../../../../").toURI().path
+    val evoSuiteTestOutput = File("$classPath/evosuite-tests/")
+
+    fun deleteTestOutput() {
```
Is it possible to write it as
```kotlin
fun deleteTestOutput() = if (evoSuiteTestOutput.exists()) evoSuiteTestOutput.deleteRecursively()
```
? 😄

gandreadis replied:
That's not valid syntax: `If must have both main and else branches if used as an expression`

casperboone replied:
So sad.

casperboone commented on subprojects/test-generator/src/test/kotlin/org/cafejojo/schaapi/testgenerator/EvoSuiteRunnerTest.kt:
```diff
@@ -0,0 +1,40 @@
+                "org.cafejojo.schaapi.test.EvoSuiteTestClass",
+                classPath,
+                classPath,
+                    generationTimeoutSeconds = 5
```
The indentation looks off here.

Also, why is this one 'named' while the others aren't? :)

gandreadis replied:
Because this one has a default, and the other don't :) But if you want, I can add the names to the others, as well (not sure what the common practice is here)

casperboone commented on subprojects/test-generator/src/main/kotlin/org/cafejojo/schaapi/testgenerator/EvoSuiteRunner.kt:
```diff
@@ -0,0 +1,69 @@
+        receiveProcessOutput(process)
+    }
+
+    private fun buildProcess(): Process {
```
```kotlin
private fun buildProcess() = ProcessBuilder(...).start()
```
? 😄

casperboone commented on subprojects/test-generator/src/main/kotlin/org/cafejojo/schaapi/testgenerator/EvoSuiteRunner.kt:
```diff
@@ -0,0 +1,69 @@
+    /**
+     * Runs the EvoSuite test generator in a new process.
+     */
+    fun run() {
```
```kotlin
fun run() = receiveProcessOutput(buildProcess())
```
? 😄

casperboone commented on subprojects/test-generator/gradle.properties:
```diff
@@ -0,0 +1,2 @@
+sootVersion=3.1.0
```
Could you align the version numbers?

casperboone replied:
Oh and actually, I think the Soot version can be removed

casperboone commented on subprojects/test-generator/build.gradle:
```diff
@@ -0,0 +1,7 @@
+apply plugin: "application"
+
+mainClassName = "org.cafejojo.schaapi.testgenerator.TestGenerator"
```
This class does not seem to actually exist.

gandreadis replied:
~~Out of scope for this PR~~ I'll change it then for now, but not really relevant at this point

gandreadis commented on subprojects/test-generator/gradle.properties:
```diff
@@ -0,0 +1,2 @@
+sootVersion=3.1.0
```
That was the autoformatting acting up, I'll fix that 👍

gandreadis commented on subprojects/test-generator/src/main/kotlin/org/cafejojo/schaapi/testgenerator/EvoSuiteRunner.kt:
```diff
@@ -0,0 +1,69 @@
+        private val classPath: String,
+        private val outputDirectory: String,
+        private val generationTimeoutSeconds: Int = 60,
+        private val logEvoSuiteOutput: Boolean = false
```
Good suggestion!

gandreadis commented:
@FWDekker I thought about adding other tests, but I can't really see what I could meaningfully test further. Inspecting the contents of the file, for example, is risky, as EvoSuite's behavior (and thus output) is stochastic. But do let me know if you have suggestions for tests

FWDekker commented:
@gandreadis Maybe test that errors are thrown if the classpath is invalid, or if the target class does not exist, etc.? I think we should verify that we can properly detect when Evo returns non-zero.

gandreadis commented:
@FWDekker Good suggestion. The problem: EvoSuite doesn't return non-zero in those cases. It just outputs a help message (different for each case). Do we want to query this help message string and based on that throw other exceptions?

casperboone commented:
@gandreadis I think that would be a good idea, yes. Just throw an exception for everything that's not the success message I'd say.

JSAbrahams commented on subprojects/test-generator/src/main/kotlin/org/cafejojo/schaapi/testgenerator/EvoSuiteRunner.kt:
```diff
@@ -0,0 +1,63 @@
+ * @property classPath the class path on which to find the class that should be tested
+ * @property outputDirectory the output directory path for the generated EvoSuite tests
+ * @property generationTimeoutSeconds how long to let the EvoSuite test generator run (in seconds)
+ * @property logEvoSuiteOutput whether to output EvoSuite logs to [System.out]
```
Should probably be changed to `evoSuiteOutputStream`

gandreadis replied:
Good catch, will fix

JSAbrahams commented on subprojects/test-generator/build.gradle:
```diff
@@ -0,0 +1,7 @@
+apply plugin: "application"
+
+mainClassName = "org.cafejojo.schaapi.testgenerator.TestGeneratorKt"
```
Not sure if there is a consensus on this yet, but I prefer it without `Kt` 😅

gandreadis replied:
The other files have this, as well, so for consistency's sake, I suggest we keep it with the `Kt` suffix

casperboone replied:
I believe that the suffix is non optional

gandreadis replied:
I even believe it is necessary, as there is no class `TestGenerator` - the `main` function is on file-level, so likely it is compiled into a class `TestGeneratorKt`

JSAbrahams replied:
Hmmm most seem to have it except for the `project-compiler` package, so it isnt necessary it seems. But since the majority uses `Kt`, and because it's not really important it should probably stay as it is.

JSAbrahams commented on subprojects/test-generator/src/main/kotlin/org/cafejojo/schaapi/testgenerator/EvoSuiteRunner.kt:
```diff
@@ -0,0 +1,63 @@
+        while (line != null) {
+            evoSuiteOutputStream?.write(line.toByteArray())
+
+            line = bufferedReader.readLine()
```
You forget to close the `bufferedReader` I believe. But you could alternatively use `forEachLine`:

`bufferedReader.forEachLine { evoSuiteOutputStream?.write(it.toByteArray()) }`

Which also closes the `BufferedReader` for you.

gandreadis replied:
Nice! Didn't know that one, definitely more elegant :)

gandreadis commented:
@JSAbrahams We could, but it would boil down to us writing it to a file and then having it read by EvoSuite (the commandline tool doesn't take a string format for bytecode - that needs to be a file). Perhaps something that can be added in a later PR, though

gandreadis commented:
@FWDekker @casperboone I've added two tests, and implemented a check for EvoSuite's last output line (which indicates success).

@JSAbrahams I've implemented your suggestions, except for the generated byte code feature (which I believe to be out of scope for this PR). Let me know if you have any follow-ups!

casperboone commented on subprojects/test-generator/src/main/kotlin/org/cafejojo/schaapi/testgenerator/EvoSuiteRunner.kt:
```diff
@@ -0,0 +1,68 @@
+        process.waitFor()
+
+        if (!lastLine.toLowerCase().contains("computation finished")) {
+            throw EvoSuiteRuntimeException("EvoSuite did not terminate successfully.")
```
Can we include the last line in the message?

gandreadis replied:
Good suggestion, fixed now

FWDekker commented on subprojects/test-generator/src/main/kotlin/org/cafejojo/schaapi/testgenerator/EvoSuiteRunner.kt:
```diff
@@ -0,0 +1,69 @@
+ * @property classPath the class path on which to find the class that should be tested
+ * @property outputDirectory the output directory path for the generated EvoSuite tests
+ * @property generationTimeoutSeconds how long to let the EvoSuite test generator run (in seconds)
+ * @property evoSuitePrintStream whether to output EvoSuite logs to [System.out]
```
The documentation of the parameter still describes a boolean.

FWDekker replied:
Am I correct that it's the caller's responsibility to close the stream?

gandreadis commented:
@FWDekker Good points, fixed in the latest commit
